projects/com.oracle.truffle.llvm.libraries.bitcode/src/crt0.c

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of
+ * conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdlib.h>
+
+int main(int argc, char **argv, char **envp);
```

Comment:
This does not match the signature of the Rust main function, which takes an i64 as argument count. A `ClassCastException` will be thrown on the attempt to pass the integer argc for the long argc to the Rust main function. Maybe choosing `_start` as entry point by default is overall too C specific. Could you generalize your code in a way that it will not break support for other llvm languages like Rust? - for example, using the old implementation as fallback if the signature of the main function does not match.

Reply:
Basically `_start` has nothing to do with C at all. It's the entry point of every executable on Linux. The value of `_start` is used for the `e_entry` field in the ELF header. If no `_start` is found, the linker prints a warning and defaults to the platform's default start address. On x86_64 this might look like this:
```
/usr/bin/ld: warning: cannot find entry symbol _start; defaulting to 0000000000000310
```
In C, the `_start` label is [defined in the libc](https://git.musl-libc.org/cgit/musl/tree/arch/x86_64/crt_arch.h). The `_start` code in the libc normally sets up `argc` and `argv` and [then calls some entry point of the libc](https://git.musl-libc.org/cgit/musl/tree/crt/crt1.c), which then [initializes various things (including TLS) and then calls `main`](https://git.musl-libc.org/cgit/musl/tree/src/env/__libc_start_main.c).

To be precise: `_start` is not even a function. It is just a label where execution starts. The single "argument" in this bitcode implementation is normally retrieved by assembler code, because it's the stack pointer when execution starts at `_start`. On Linux the stack pointer points to `argc`, followed by `argv`, followed by `env`, followed by `auxv`, followed by auxv data, followed by argv strings, followed by env strings (followed by the execfn string). Therefore, you cannot `return` from `_start` (because there is no return address) and you have to call the `SYS_exit` syscall to terminate the program.

`main` *is* C specific and defined in the C standard. That's why you can define your own `_start` if you don't use a libc. That's what languages like Rust do in their runtime libraries.

The only question is how to distinguish languages which use a C style `main` from other languages like Rust which don't use a C style `main`.

Reply:
But once it is compiled to llvm IR, `_start` ends up to be just an ordinary function, right? Anyways, it shouldn't be a problem if your implementation of `_start` became the entry point for every llvm language executed on top of Sulong. The problem is just the signature of the main function. Therefore, I would say, either we have a `_start` that can handle any main function signature specified in the bitcode that gets executed or we use the old approach which calls the main function directly as fallback in the case that Sulong encounters a non C style main.